eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/TxPublisher.scala

```diff
@@ -0,0 +1,272 @@
+          val pending2 = if (remainingAttempts.isEmpty) pending - cmd.input else pending + (cmd.input -> remainingAttempts)
+          reason match {
+            case WalletInputGone =>
+              // Our transaction has been evicted from the mempool because it depended on an unconfirmed input that has
+              // been replaced. We should be able to retry right now with new wallet inputs (no need to wait for a new
+              // block).
+              timers.startSingleTimer(cmd, (1 + Random.nextLong(nodeParams.maxTxPublishRetryDelay.toMillis)).millis)
+              run(pending2, retryNextBlock, channelInfo)
+            case CouldNotFund =>
+              // We don't have enough funds at the moment to afford our target feerate, but it may change once pending
+              // transactions confirm, so we retry when a new block is found.
+              run(pending2, retryNextBlock ++ rejectedAttempts.map(_.cmd), channelInfo)
+            case TxSkipped(retry) =>
+              val retryNextBlock2 = if (retry) retryNextBlock ++ rejectedAttempts.map(_.cmd) else retryNextBlock
+              run(pending2, retryNextBlock2, channelInfo)
+            case ConflictingTxUnconfirmed =>
+              // Our transaction was replaced by a transaction that pays more fees, so it doesn't make sense to retry now.
+              // We will automatically retry with a higher fee if we get close to the deadline.
+              run(pending2, retryNextBlock, channelInfo)
+            case ConflictingTxConfirmed =>
+              // Our transaction was double-spent by a competing transaction that has been confirmed, so it doesn't make
+              // sense to retry.
+              run(pending2, retryNextBlock, channelInfo)
+            case UnknownTxFailure =>
+              // We don't automatically retry unknown failures, they should be investigated manually.
+              run(pending2, retryNextBlock, channelInfo)
+          }
+      }
+
+      case WrappedCurrentBlockCount(currentBlockCount) =>
+        log.debug("retry publishing {} transactions at block {}", retryNextBlock.length, currentBlockCount)
+        retryNextBlock.foreach(cmd => timers.startSingleTimer(cmd, (1 + Random.nextLong(nodeParams.maxTxPublishRetryDelay.toMillis)).millis))
+        run(pending, Seq.empty, channelInfo)
+
+      case SetChannelId(remoteNodeId, channelId) =>
+        Behaviors.withMdc(Logs.mdc(remoteNodeId_opt = Some(remoteNodeId), channelId_opt = Some(channelId))) {
+          run(pending, retryNextBlock, channelInfo.copy(remoteNodeId = remoteNodeId, channelId_opt = Some(channelId)))
+        }
+    }
+  }
+
+  private def stopAttempts(attempts: Seq[PublishAttempt]): Unit = attempts.foreach(stopAttempt)
+
+  private def stopAttempt(attempt: PublishAttempt): Unit = attempt match {
+    case RawAttempt(_, _, actor) => actor ! RawTxPublisher.Stop
+    case ReplaceableAttempt(_, _, _, actor) => actor ! ReplaceableTxPublisher.Stop
+  }
```

Comment:
I don't have a strong opinion on this, but I liked defining the `Attempt` trait inside this private class since it's an internal implementation detail. If we move this function in the `object`, we also need to move the `Attempt` trait...not sure if it's good or not.

Reply:
Fair enough